[Doc] Add callout for security domain usage (#87408)
elastic/elasticsearch: x-pack/docs/en/security/authentication/security-domain.asciidoc

API keys. If that user doesn't have this privilege when authenticating with one
realm, they won't be able to manage API keys while authenticating with another
realm.

[[security-domain-realm-roles]]
===== Managing roles across realms
{es} provides multiple ways to consistently apply roles across realms. For example, you can use
<<configuring-authorization-delegation,authorization delegation>> to
ensure that a user is assigned the same roles from multiple realms. You can also
manually configure multiple realms that are backed by the same directory service.
Though it's possible to configure different <<roles,roles>> for the same user
when authenticating with different realms, it is _not_ recommended.

[[security-domain-warning]]]]
[IMPORTANT]
====
Security domain is an advanced feature and must be used with care. It can lead
to unexpected behaviours if incorrectly configured or misused.
Here are few things watch out when working with security domain:

* Security domains must be configured consistently across all nodes. Since changing
security domain requires node restart, this means they may not be fully functional

> during rolling restart.

Reviewer: Can we be more specific here?
As it reads, it says "be careful" and "rolling restarts might not work well", but doesn't tell the admin how to avoid potential problems. Do we just expect them not to do any changes? Or do full cluster restarts?

It would be nice to tell them something that they can actually respond to (if that's possible).

Author: I expanded a bit (dos and donts) on this.

> * Security domains are better at grouping multiple realms, but not so much at splitting them afterwards.

Reviewer: Maybe something like

```suggestion
* Security domains are designed to be effective for grouping multiple realms. Splitting existing
realms out of a security domain is more complex and can introduce unexpected behaviors.
```

This means assigning realms under a domain works as expected (assuming they are configured

> consistently across all nodes).
> But removing realms from a domain must be proceeded with caution. Resources

Reviewer: ```suggestion
consistently across all nodes), but removing realms from a domain must be
performed with caution. Resources
```

created before domain splitting can be owned by different users depending on resource types:

>   - <<user-profile,User profile>> will be owned by the user who performs the last
> <<security-api-activate-user-profile,Activate User Profile>> request.

Reviewer: ```suggestion
  - <<user-profile,User profiles>> will be owned by the user for whom it was last
<<security-api-activate-user-profile,activated>>.
```

If the owner user's realm is under a domain, the profile is shared across realms in that
domain for the user. Other users that are no longer in the same domain as the owner user will get
new profiles created the next time they call the Activate User Profile API.
  - Other resources, such as API keys, will be owned by the user who originally created them.
If the owner user's realm is under a domain, the resource is shared across realms in that
domain for the user.

NOTE: The "same username" rule is always respected. That is, resource sharing across realms
is only possible between users with the same username under all circumstances.

> ====

Author: This is the only real change. Other places are due to formatting issues.

realms out of a security domain is more complex and can introduce unexpected behaviors.
This means assigning realms under a domain works as expected (assuming they are configured
consistently across all nodes), but removing realms from a domain must be
performed with caution. Resources
created before domain splitting can be owned by different users depending on resource types:

- <<user-profile,User profiles>> will be owned by the user for whom it was last
<<security-api-activate-user-profile,activated>>.
If the owner user's realm is under a domain, the profile is shared across realms in that
domain for the user. Other users that are no longer in the same domain as the owner user will get
new profiles created the next time they call the Activate User Profile API.
- Other resources, such as API keys, will be owned by the user who originally created them.
If the owner user's realm is under a domain, the resource is shared across realms in that
domain for the user.

NOTE: The "same username" rule is always respected. That is, resource sharing across realms
is only possible between users with the same username under all circumstances.

> ====

Reviewer 2: I think this is too much to try put in an admonition. I'd suggest incorporating most of it into the Configure a security domain section, and adding a section with the info about removing realms being problematic. See https://github.com/elastic/elasticsearch/commit/23710d55ba8fc5c3be1abe2d8f9abaa7bdf74f90.

Author: Thanks Deb. The changes look good to me. They retain the same central idea but present it in a much better format. Do you mind directly pushing to this PR? Alternatively, I can incorporate them on my side. Please let me know,.
